enclave/sgx/collateralinfo.c

```diff
@@ -32,6 +32,7 @@ oe_result_t oe_get_sgx_quote_verification_collateral(
     for (;;)
     {
         memcpy(&out, &in, sizeof(out));
+        oe_prealloc_quote_verification_collateral_args(&out);
```

Comment:
Just call prealloc once before entering the loop

Reply:
The memcpy in line 34 would probably erase the pre-allocation if I called it outside the for loop. The 'in' buffer has all its args except fmspc, pointing to NULL. I can add a condition to execute this function only once, if we think that can save us some runtime.

Reply:
I'm not too worried about the runtime overhead, I was just thinking it could simplify the code in `oe_prealloc_quote_verification_collateral_args`. This isnt too big of a deal though.

Reply:
It's a valid concern. If the initial buffer is too small, line 34 will set the expected buffer sizes to 'out' variable. Calling line 35 again after the first iteration will wipe out the purpose of line 34. This may lead to infinite looping.

---
In reply to: [418278362](https://github.com/openenclave/openenclave/pull/2969#discussion_r418278362) [](ancestors = 418278362)

Reply:
Oh @shruti25ratnam now that I think about it, you could just call `oe_prealloc_quote_verification_collateral_args` on `&in` before the loop. Then the pointers will get copied to out

Reply:
Please correct me if I'm wrong, per convention I thought 'in' buffer represents the values coming in from the ocall and 'out' represented variables passed out to the ocall. I'll go per your recommendation if this is not the case.

Reply:
> It's a valid concern. If the initial buffer is too small, line 34 will set the expected buffer sizes to 'out' variable. Calling line 35 again after the first iteration will wipe out the purpose of line 34. This may lead to infinite looping.
> 
> In reply to: [418278362](https://github.com/openenclave/openenclave/pull/2969#discussion_r418278362) [](ancestors = 418278362)

In the function in line 35, I only pre-allocate if the sizes of the collateral args are zero, which is usually only in the first iteration. If sizes are non-zero this function just passes through. This saves it from infinite looping. 

Although, I'm currently adding some changes to use the second iteration to replace the estimated sizes with the actual ones and making them static. So that the eventual calls to this function could use the actual values for pre-allocation.

Reply:
> per convention I thought 'in' buffer represents the values coming in from the ocall and 'out' represented variables passed out to the ocall

You're basically using the values returned from the ocall as the "input" the next ocall. It just so happens that you already have some input values you want to try first. I would say it's fine.

> In the function in line 35, I only pre-allocate if the sizes of the collateral args are zero, which is usually only in the first iteration. If sizes are non-zero this function just passes through. This saves it from infinite looping.

Yes that's true. I just think the code in the function could be a lot simpler if you don't have to check for pre-populated sizes.

Reply:
Sounds good to me. 
@yentsanglee do you agree with the suggestion of pulling the function outside the loop? Upside: code is much simpler. Downside: we may not be able to use static variables to store the actual collateral sizes retrieved from the first run, as you suggested earlier.

Reply:
You can still use static values right? Just preallocate outside the loop using the static values and set them inside the loop. The next call to this function will use the updated values

Reply:
I think Shruti's plan was to keep those static variables in oe_prealloc_quote_verification_collateral_args(). However, the function name oe_prealloc_quote_verification_collateral_args() implies that it is to be called once. In order for it to also store previous values, its purpose becomes ambiguous. Jordan is suggesting declaring static values in oe_get_sgx_quote_verification_collateral() and passing them in to 'pre-allocate' and 'store' functions separately and that makes sense to me.

Reply:
Sure I am making these changes. Thanks for the suggestions.